packages/firestore/test/integration/api/database.test.ts

```diff
@@ -514,4 +515,55 @@ apiDescribe('Database', persistence => {
       return Promise.resolve();
     });
   });
+
+  asyncIt('can queue writes while offline', () => {
+    return withTestDb(persistence, db => {
+      const docRef = db.collection('rooms').doc();
+      const firestoreClient = (docRef.firestore as Firestore)._firestoreClient;
+
+      return firestoreClient
+        .disableNetwork()
+        .then(() => {
+          return Promise.all([
+            docRef.delete(),
+            firestoreClient.enableNetwork()
+          ]);
+        })
+        .then(() => Promise.resolve());
+    });
+  });
+
+  asyncIt('can get documents while offline', () => {
+    return withTestDb(persistence, db => {
+      const docRef = db.collection('rooms').doc();
+      const firestoreClient = (docRef.firestore as Firestore)._firestoreClient;
+
+      return firestoreClient
+        .disableNetwork()
+        .then(() => {
+          const promises = [];
+          let done: () => void;
+
+          promises.push(docRef.set({ a: 1 }));
+          promises.push(
+            new Promise(resolve => {
+              done = resolve;
+            })
+          );
+
+          docRef.get().then(snapshot => {
+            expect(snapshot.metadata.fromCache).to.be.true;
+            firestoreClient.enableNetwork().then(() => {
+              return docRef.get().then(snapshot => {
+                expect(snapshot.metadata.fromCache).to.be.false;
+                done();
```

Comment:
Actually, I'm fuzzy why you need the done thing at all.  Why can't you just promises.push() this whole docRef.get().then(snapshot => ...) block?  (you probably need to "return firestoreClient.enableNetwork()..." a couple lines up too)

Reply:
That is indeed a great simplification. And it turns out that I don't even need to keep all promises around, I just need a reference to the first promise.